### DISCUSSION TITLE
Bootstrap on remote board

### DISCUSSION THREAD

**Discussion by Mdfayaz003** [labgrid-project/labgrid#1103] [General]
I tried the bootstrap command from client and found it uses ssh in backend.
**Command gave:**
```
labgrid-client -p board-sx -c ../config-labgrid/client-1.yaml -d bootstrap /path/to/u-boot-sd.imx
```
I got a error as Permission denied:
```
Selected role imx6sx from configuration file
/home/jidtp1401/Labgrid/.labgrid-venv/lib/python3.10/site-packages/labgrid/factory.py:137: UserWarning: unsupported attribute 'cmd' with value 'spl' for class 'UUUDriver'
  args = filter_dict(args, cls, warn=True)
Traceback (most recent call last):
  File "/home/jidtp1401/Labgrid/.labgrid-venv/lib/python3.10/site-packages/labgrid/remote/client.py", line 1787, in main
    args.func(session)
  File "/home/jidtp1401/Labgrid/.labgrid-venv/lib/python3.10/site-packages/labgrid/remote/client.py", line 930, in bootstrap
    drv.load(self.args.filename)
  File "/home/jidtp1401/Labgrid/.labgrid-venv/lib/python3.10/site-packages/labgrid/binding.py", line 96, in wrapper
    return func(self, *_args, **_kwargs)
  File "/home/jidtp1401/Labgrid/.labgrid-venv/lib/python3.10/site-packages/labgrid/step.py", line 215, in wrapper
    _result = func(*_args, **_kwargs)
  File "/home/jidtp1401/Labgrid/.labgrid-venv/lib/python3.10/site-packages/labgrid/driver/usbloader.py", line 78, in load
    mf.sync_to_resource()
  File "/home/jidtp1401/Labgrid/.labgrid-venv/lib/python3.10/site-packages/labgrid/util/managedfile.py", line 56, in sync_to_resource
    conn = sshmanager.open(host)
  File "/home/jidtp1401/Labgrid/.labgrid-venv/lib/python3.10/site-packages/labgrid/util/ssh.py", line 76, in open
    return self.get(host)
  File "/home/jidtp1401/Labgrid/.labgrid-venv/lib/python3.10/site-packages/labgrid/util/ssh.py", line 50, in get
    instance.connect()
  File "/home/jidtp1401/Labgrid/.labgrid-venv/lib/python3.10/site-packages/labgrid/util/ssh.py", line 399, in connect
    self._open_connection()
  File "/home/jidtp1401/Labgrid/.labgrid-venv/lib/python3.10/site-packages/labgrid/util/ssh.py", line 174, in _open_connection
    self._start_own_master()
  File "/home/jidtp1401/Labgrid/.labgrid-venv/lib/python3.10/site-packages/labgrid/util/ssh.py", line 457, in _start_own_master
    raise ExecutionError(
labgrid.driver.exception.ExecutionError: failed to connect to 172.21.3.87 with args ['ssh', '-x', '-o', 'LogLevel=ERROR', '-o', 'PasswordAuthentication=no', '-n', '-MN', '-o', 'ConnectTimeout=30', '-o', 'ControlPersist=300', '-o', 'ControlMaster=yes', '-o', 'ControlPath=/tmp/lg-con-np_9pvah/control-172.21.3.87', '-o', 'StrictHostKeyChecking=yes', '172.21.3.87'], returncode=255 b'',b'jidtp1401@172.21.3.87: Permission denied (publickey).\r\n'
```
I tried the same command given at backend:
```
ssh -x -o LogLevel=ERROR -o PasswordAuthentication=no -n -MN -o ConnectTimeout=30 -o ControlPersist=300 -o ControlMaster=yes  -o StrictHostKeyChecking=no 172.21.3.87
```
And resulted as: `Permission denied (publickey).`

After then I tried:
```
ssh -x -o LogLevel=ERROR -o PasswordAuthentication=yes -n -MN -o ConnectTimeout=30 -o ControlPersist=300 -o ControlMaster=yes  -o StrictHostKeyChecking=no 172.21.3.87
```
And then asked for my PC password, then nothing was printed

Is there any option to change that parameter or any other solutions?

**Comment by Emantor** [comment]
Labgrid requires SSH publickey authentication, so you'll need to setup that first and ensure you can login on the remote host without typing your password.

**Reply by Mdfayaz003** [reply]
Thanks @Emantor for your instant replay.
I may try it and update you

**Reply by Mdfayaz003** [reply]
Thanks, @Emantor, It worked after doing the necessary steps of SSH public-key authentication.

**Configuration of client:**
```
targets:
  imx6sx:
    resources:
      RemotePlace:
        name: board-sx
    drivers:
      ManualPowerDriver:
        name: manual
      SerialDriver: {}
      SSHDriver: {}
      IMXUSBDriver:
        image: full_image
      UUUDriver:
        image: full_image
images:
  full_image: /path/to/u-boot-sd.imx
tools:
  imx-usb-loader: /path/to/imx-usb
```
**Configuration of expoter:**
```
group-1:
  IMXUSBLoader:
    match:
      ID_PATH: pci-0000:00:14.0-usb-0:12
  USBSerialPort:
    match:
      ID_PATH: pci-0000:00:14.0-usb-0:10:1.0
    speed: 115200
  NetworkService:
    address: 172.21.3.5
    username: mdfayaz
    password: "********"
```
Then tried the command `bootstrap`  with debug mode enabled and found something different from what I thought.

I need some clarification on the configuration file used on the client side.
1. What binary file should be input for `imx-usb-loader`?
     I used a build binary of NXP's `imx_usb` which was built locally. I found that the options available on that binary and the options are given by labgrid mismatch
    Can you share the proper binary to be used for it?

2. Is it possible to use NXP's UUU tool for flashing binaries?
3. How to use UUUDriver for flashing binaries?
4. What are the options available for `tools` keywords like `imx-usb-loader`?
5. Is the format which is given for `images` correct?

**Reply by Mdfayaz003** [reply]
Hi all, I figured out how to use NXP's UUU tool for flashing binaries.

**Configuration of client:**
```
targets:
  imx6sx:
    resources:
      RemotePlace:
        name: board-sx
    drivers:
      ManualPowerDriver:
        name: manual
      SerialDriver: {}
      SSHDriver: {}
      UUUDriver:
        script: /path/to/flash-script.uuu
        cmd: spl
tools:
  uuu-loader: /path/to/uuu
```
**Configuration of expoter:**
```
group-1:
  IMXUSBLoader:
    match:
      ID_PATH: pci-0000:00:14.0-usb-0:12
  USBSerialPort:
    match:
      ID_PATH: pci-0000:00:14.0-usb-0:10:1.0
    speed: 115200
  NetworkService:
    address: 172.21.3.5
    username: mdfayaz
    password: "********"
```
Then tried `bootstrap` command and it worked as expected.

On trying to access the target board remotely, I run the coordinator and exporter on Ubuntu 18.04 (board connected) and client on Ubuntu 22.04. Used the same configuration files with same changes respective to PC's.

I got `Permission Denied Error` since it was trying to create a series of files on location `/var/cache/`

I need some explanation why it need to access `/var/cache/` and a solution to solve this problem

**Reply by Bastian-Krause** [reply]
Please don't use the same discussion for multiple questions that don't have anything to do with each other. You are also expected to at least search the documentation (e.g. `/var/cache`, `images`) and have a look at the code (e.g. UUUDriver) before asking. If you spent some time investigating and you can't come up with a solution, you're welcome to open a discussion per topic with a _specific_ question.
